File: 62 Mugwump/javascript/mugwump.js
// MUGWUMP
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

var p = [];

// Main program
async function main()
{
    print(tab(33) + "MUGWUMP\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    // Courtesy People's Computer Company
    print("THE OBJECT OF THIS GAME IS TO FIND FOUR MUGWUMPS\n");
    print("HIDDEN ON A 10 BY 10 GRID.  HOMEBASE IS POSITION 0,0.\n");
    print("ANY GUESS YOU MAKE MUST BE TWO NUMBERS WITH EACH\n");
    print("NUMBER BETWEEN 0 AND 9, INCLUSIVE.  FIRST NUMBER\n");
    print("IS DISTANCE TO RIGHT OF HOMEBASE AND SECOND NUMBER\n");
    print("IS DISTANCE ABOVE HOMEBASE.\n");
    print("\n");
    print("YOU GET 10 TRIES.  AFTER EACH TRY, I WILL TELL\n");
    print("YOU HOW FAR YOU ARE FROM EACH MUGWUMP.\n");
    print("\n");
    while (1) {
        for (i = 1; i <= 4; i++) {
            p[i] = [];
            for (j = 1; j <= 2; j++) {
                p[i][j] = Math.floor(10 * Math.random());
            }
        }
        t = 0;
        do {
            t++;
            print("\n");
            print("\n");
            print("TURN NO. " + t + " -- WHAT IS YOUR GUESS");
            str = await input();
            m = parseInt(str);
            n = parseInt(str.substr(str.indexOf(",") + 1));
            for (i = 1; i <= 4; i++) {
                if (p[i][1] == -1)
                    continue;
                if (p[i][1] == m && p[i][2] == n) {
                    p[i][1] = -1;
                    print("YOU HAVE FOUND MUGWUMP " + i + "\n");
                } else {
                    d = Math.sqrt(Math.pow(p[i][1] - m, 2) + Math.pow(p[i][2] - n, 2));
                    print("YOU ARE " + Math.floor(d * 10) / 10 + " UNITS FROM MUGWUMP " + i + "\n");
                }
            }
            for (j = 1; j <= 4; j++) {
                if (p[j][1] != -1)
                    break;
            }
            if (j > 4) {
                print("\n");
                print("YOU GOT THEM ALL IN " + t + " TURNS!\n");
                break;
            }
        } while (t < 10) ;
        if (t == 10) {
            print("\n");
            print("SORRY, THAT'S 10 TRIES.  HERE IS WHERE THEY'RE HIDING:\n");
            for (i = 1; i <= 4; i++) {
                if (p[i][1] != -1)
                    print("MUGWUMP " + i + " IS AT (" + p[i][1] + "," + p[i][2] + ")\n");
            }
        }
        print("\n");
        print("THAT WAS FUN! LET'S PLAY AGAIN.......\n");
        print("FOUR MORE MUGWUMPS ARE NOW IN HIDING.\n");
    }
}

main();

File: 62_Mugwump/perl/mugwump.pl
#!/usr/bin/perl

use strict;
use warnings;

# global variables defined here
my(@MUGWUMP) = ();

# subroutines defined here

# init_mugwump: pick the random places for the Mugwumps
sub init_mugwump() {
    @MUGWUMP = (); 
    for (1 .. 4) {
        push @MUGWUMP, [ int(rand 10), int(rand 10) ];
    }
}


# main code starts here

# print introductory text
print <<HERE;
                  Mugwump
Creative Computing  Morristown, New Jersey



The object of this game is to find four Mugwumps
hidden on a 10 by 10 grid.  Homebase is position 0,0.
Any guess you make must be two numbers with each
number between 0 and 9, inclusive.  First number
is distance to right of homebase and second number
is distance above homebase.

You get 10 tries.  After each try, I will tell
you how far you are from each Mugwump.

HERE


# PLAY block implements a complete game, and the
# continue block prints the "let's play again" msg
PLAY: while (1) {

    init_mugwump();
    TURN: for my $turn (1 .. 10) { 

        printf("\nTurn no %d -- what is your guess ? ", $turn);

        # Note that parsing of input is rudimentary, in keeping with the
        # spirit of the original BASIC program. Increased input checks
        # would be a good place to start working on this program!
        chomp(my $in = <STDIN>);
        my($M,$N) = split(/,/,$in);
        $M = int($M);
        $N = int($N);

        for my $i (0 .. $#MUGWUMP) {
            # -1 indicates a Mugwump that was already found
            next if $MUGWUMP[$i]->[0] == -1;

            if ($MUGWUMP[$i]->[0] == $M && $MUGWUMP[$i]->[1] == $N) {
                $MUGWUMP[$i]->[0] = -1;
                printf("You have found Mugwump %d\n", $i+1);
            } else {
                my $d = sqrt(($MUGWUMP[$i]->[0] - $M) ** 2 + ($MUGWUMP[$i]->[1] - $N) ** 2);
                printf("You are %.1f units away from Mugwump %d\n", $d, $i+1);
            }
        }

        # If a Mugwump still has not been found,
        # go to the next turn
        for my $j (0 .. $#MUGWUMP) {
            if ($MUGWUMP[$j]->[0] != -1) {
                next TURN;
            }
        }
        # You win!
        printf("You got all of them in %d %s!\n\n", $turn, ($turn == 1 ? 'turn' : 'turns'));
        # Pass execution down to the continue block
        next PLAY;

    } # end of TURN loop

    print "\nSorry, that's 10 tries.  Here's where they're hiding:\n";
    for my $i (0 .. $#MUGWUMP) {
        printf("Mugwump %d is at (%d, %d)\n", $i+1, $MUGWUMP[$i]->[0], $MUGWUMP[$i]->[1])
            if $MUGWUMP[$i]->[0] != -1;
    }
} 
continue {
    print "\nThat was fun! Let's play again.......\n";
    print "Four more Mugwumps are now in hiding.\n\n";
}

